### PR TITLE
chore(prometheus_scrape source, prometheus_remote_write source, prometheus_exporter sink, prometheus_remote_write sink): add tests for duplicate labels

### LINKED ISSUE #15516: `prometheus_remote_write` enhanced tag support

### LINKED ISSUE #15517: `prometheus_scrape` enhanced tag support

### LINKED ISSUE #15581: `prometheus_exporter` enhanced tag support

### LINKED ISSUE #15582: `prometheus_remote_write` enhanced tag support

### PR DESCRIPTION
**Pull request by StephenWakely**
Closes #15581
Closes #15582 
Epic #15420 

Closes #15516 
Closes #15517 
Epic #15419 

Prometheus considers duplicate tags to be invalid. Sending a metric to Prometheus with a duplicate tag results in an error. 
Vector doesn't reject the error, but will take the last tag sent. Largely to maintain backward compatibility with current versions of Vector we have decided to maintain this behaviour. This PR adds some tests to codify it.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: src/sinks/prometheus/collector.rs [hunk 1]
```diff
@@ -938,4 +938,31 @@ mod tests {
             "#}
         );
     }
+
+    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
+    /// > Label names MUST be unique within a LabelSet.
+    /// Prometheus itself will reject the metrict with an error. Largely to remain backward compatible with older versions of Vecotr,
+    /// we only pushlish the last tag in the list.
```
Comment (spencergilbert) [inline src/sinks/prometheus/collector.rs:945]:
```suggestion
    /// Prometheus itself will reject the metric with an error. Largely to remain backward compatible with older versions of Vector,
    /// we only take the last tag in the list.
```

File: src/sinks/prometheus/exporter.rs [hunk 2]
```diff
@@ -828,6 +831,35 @@ mod tests {
         );
     }
 
+    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
+    /// > Label names MUST be unique within a LabelSet.
+    /// Prometheus itself will reject the metrict with an error. Largely to remain backward compatible with older versions of Vecotr,
+    /// we only pushlish the last tag in the list.
```
Comment (spencergilbert) [inline src/sinks/prometheus/exporter.rs:837]:
```suggestion
    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
    /// > Label names MUST be unique within a LabelSet.
    /// Prometheus itself will reject the metric with an error. Largely to remain backward compatible with older versions of Vector,
    /// we only take the last tag in the list.
```

File: src/sources/prometheus/remote_write.rs [hunk 3]
```diff
@@ -261,6 +264,77 @@ mod test {
             .into(),
         ]
     }
+
+    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
+    /// > Label names MUST be unique within a LabelSet.
+    /// Prometheus itself will reject the metrict with an error. Largely to remain backward compatible with older versions of Vecotr,
+    /// we accept the metric, but take the last label in the list.
```
Comment (spencergilbert) [inline src/sources/prometheus/remote_write.rs:271]:
```suggestion
    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
    /// > Label names MUST be unique within a LabelSet.
    /// Prometheus itself will reject the metric with an error. Largely to remain backward compatible with older versions of Vector,
    /// we accept the metric, but take the last label in the list.
```

File: src/sources/prometheus/scrape.rs [hunk 4]
```diff
@@ -458,6 +458,62 @@ mod test {
         }
     }
 
+    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
+    /// > Label names MUST be unique within a LabelSet.
+    /// Prometheus itself will reject the metrict with an error. Largely to remain backward compatible with older versions of Vecotr,
+    /// we accept the metric, but take the last label in the list.
```
Comment (spencergilbert) [inline src/sources/prometheus/scrape.rs:464]:
```suggestion
    /// According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L115)
    /// > Label names MUST be unique within a LabelSet.
    /// Prometheus itself will reject the metric with an error. Largely to remain backward compatible with older versions of Vector,
    /// we accept the metric, but take the last label in the list.
```

**Review by spencergilbert — CHANGES_REQUESTED** [review]

File: src/sinks/prometheus/collector.rs (hunk 1 above)
Comment (StephenWakely) [reply src/sinks/prometheus/collector.rs:945]: Wow I really can't tpye today..

Comment (StephenWakely) [reply src/sinks/prometheus/collector.rs:945]: Sinks publish tags, not take them - so I think we should keep publish..

Comment (spencergilbert) [reply src/sinks/prometheus/collector.rs]: Yeah, I had it `publish` originally (assumed that's what "pushlist" was) but kept editing to make things the same

**Review by spencergilbert — APPROVED** [review]

File: src/sources/prometheus/scrape.rs [hunk 5]
```diff
@@ -458,6 +458,62 @@ mod test {
...
+    /// > Label names MUST be unique within a LabelSet.
+    /// Prometheus itself will reject the metric with an error. Largely to remain backward compatible with older versions of Vector,
+    /// we accept the metric, but take the last label in the list.
+    #[tokio::test]
+    async fn test_prometheus_duplicate_tags() {
+        let in_addr = next_addr();
+
+        let dummy_endpoint = warp::path!("metrics").map(|| {
+            r#"
+                    metric_label{code="200", code="success"} 100 1612411516789
```
Comment (bruceg) [inline src/sources/prometheus/scrape.rs:471]: Is the space here valid syntax?

File: src/sources/prometheus/scrape.rs [hunk 6]
```diff
@@ -458,6 +458,62 @@ mod test {
...
+            tls: None,
+        };
+
+        let events = run_and_assert_source_compliance(
+            config,
+            Duration::from_secs(3),
+            &HTTP_PULL_SOURCE_TAGS,
+        )
+        .await;
+        assert!(!events.is_empty());
```
Comment (bruceg) [inline src/sources/prometheus/scrape.rs:495]: It might be sufficient to write these source tests test at the level of the parser instead of doing full network setup, but this isn't wrong.

**Review by bruceg — APPROVED** [review]

File: src/sources/prometheus/scrape.rs (hunk 5 above)
Comment (StephenWakely) [reply src/sources/prometheus/scrape.rs:471]: Good catch. According to the [spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?plain=1#L348), it is not - although it seems we still parse it ok.

**Review by neuronull — APPROVED** [review]
